presto-main/src/main/java/com/facebook/presto/sql/relational/optimizer/ExpressionOptimizer.java

```diff
@@ -93,6 +101,30 @@ public RowExpression visitCall(CallExpression call, Void context)
             Signature signature = call.getSignature();
 
             if (signature.getName().equals(CAST)) {
+                if (call.getArguments().get(0) instanceof CallExpression) {
+                    // Optimization for CAST(JSON_PARSE(...) AS ARRAY/MAP)
+                    CallExpression innerCall = (CallExpression) call.getArguments().get(0);
+                    if (innerCall.getSignature().getName().equals("json_parse")) {
+                        checkState(innerCall.getType().equals(JSON));
+                        checkState(innerCall.getArguments().size() == 1);
+                        Signature jsonStringCastSignature = null;
+                        if (signature.getReturnType().getBase().equals(ARRAY)) {
+                            jsonStringCastSignature = internalScalarFunction(
```

Comment:
I personally prefer to return the `RowExpression` inside this bracket. But @haozhun  will decide this :)